File: lib/gherkin/parser/filter_listener.rb
require 'gherkin/parser/sexp'
require 'gherkin/parser/tag_expression'

module Gherkin
  module Parser
    # This class filters events based on filter criteria.
    class FilterListener
      # Creates a new instance that replays events to +listener+, filtered by +filters+,
      # a Hash that can contain:
      #
      # * <tt>:lines</tt> An Array of line numbers to filter on.
      # * <tt>:name_regexen</tt> An Array of name regexen to filter on. Matches against :feature, :scenario, :scenario_outline and :examples
      # * <tt>:tag_expression</tt> A TagExpression to filter on.
      #
      def initialize(listener, filters)
        @listener, @filters = listener, filters
        @meta_buffer = []
        @feature_buffer = []
        @scenario_buffer = []
        @examples_buffer = []
        @examples_rows_buffer = []
      end
      
      private

      def method_missing(*sexp_args)
        sexp = Sexp.new(sexp_args)

        case(sexp.event)
        when :tag
          @meta_buffer << sexp
        when :comment
          @meta_buffer << sexp
        when :feature
          @feature_buffer = @meta_buffer + @feature_buffer
          @feature_buffer << sexp
          @meta_buffer = []
        when :background
          @feature_buffer += @meta_buffer
          @feature_buffer << sexp
          @meta_buffer = []
          @table_state = :background
        when :scenario, :scenario_outline
          replay_examples_rows_buffer
          @scenario_buffer = @meta_buffer
          @scenario_buffer << sexp
          @meta_buffer = []
          @scenario_ok = line_match?(*@scenario_buffer)
          @examples_ok = false
          @table_state = :step
        when :examples
          replay_examples_rows_buffer
          @examples_buffer = @meta_buffer
          @examples_buffer << sexp
          @meta_buffer = []
          @examples_rows_buffer = []
          @examples_ok = line_match?(*@examples_buffer)
          @table_state = :examples
        when :step
          case(@table_state)
          when :background
            @feature_buffer += @meta_buffer
            @feature_buffer << sexp
            @meta_buffer = []
          else
            @scenario_buffer << sexp
            @scenario_ok ||= line_match?(*@scenario_buffer)
            @table_state = :step
          end
        when :row
          case(@table_state)
          when :examples
            unless header_row_already_buffered?
              @examples_buffer << sexp
              @examples_ok = true if line_match?(*@examples_buffer)
            else
              @examples_rows_buffer << sexp if @scenario_ok || @examples_ok || line_match?(sexp)
            end
          when :step
            @scenario_buffer << sexp
            @scenario_ok ||= line_match?(*@scenario_buffer)
          when :background
            @feature_buffer += @meta_buffer
            @feature_buffer << sexp
            @meta_buffer = []
          else
            raise "BAD STATE"
          end
        when :eof
          replay_examples_rows_buffer
          sexp.replay(@listener)
          return
        else
          super
        end

        if no_filters?
          sexp.replay(@listener)
        elsif @scenario_ok || @examples_ok
          replay_buffers
        end
      end

      def no_filters?
        @filters.values.flatten.empty?
      end

      def header_row_already_buffered?
        return false unless @examples_buffer.any?
        @examples_buffer[-1].event == :row
      end
      
      def line_match?(*sexps)
        return true if no_filters?
        sexps.detect{|sexp| sexp.filter_match?(@filters)}
      end

      def lines
        @filters[:lines] || []
      end

      def names
        @filters[:name_regexen] || []
      end

      def replay_buffers
        replay_feature_buffer
        replay_scenario_buffer
      end

      def replay_examples_rows_buffer
        if @examples_rows_buffer.any?
          replay_buffers
          (@examples_buffer + @examples_rows_buffer).each do |sexp|
            sexp.replay(@listener)
          end
          @examples_rows_buffer = []
        end
      end
      
      def replay_feature_buffer
        if @feature_buffer.any?
          @feature_buffer.each{|sexp| sexp.replay(@listener) }
          @feature_buffer = []
        end
      end
      
      def replay_scenario_buffer
        if @scenario_buffer.any?
          @scenario_buffer.each{|sexp| sexp.replay(@listener) }
          @scenario_buffer = [] 
        end
      end
    end
  end
end

File: java/src/main/java/gherkin/parser/FilterListener.java
package gherkin.parser;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import gherkin.Listener;

public class FilterListener implements Listener {
	
	private final List filters;
	private final Listener listener;
	private final List<Sexp> metaBuffer = new ArrayList<Sexp>();
	private final List<Sexp> featureBuffer = new ArrayList<Sexp>();
	private final List<Sexp> scenarioBuffer = new ArrayList<Sexp>();
	private final List<Sexp> examplesBuffer = new ArrayList<Sexp>();
	private final List<Sexp> examplesRowsBuffer = new ArrayList<Sexp>();
	
	private List featureTags = new ArrayList();
	private List scenarioTags = new ArrayList();
	private List exampleTags = new ArrayList();
	
	private boolean featureOk = false;
	private TableState tableState = TableState.STEP;
	private boolean scenarioOk = false;
	private boolean examplesOk = false;
	private IFilterMethod filterMethod;

	@SuppressWarnings("unchecked")
	public FilterListener(Listener listener, List filters) {
		this.listener = listener;
		this.filters = filters;
		detectFilters(filters);
	}

	@Override
	public void tag(String name, int line) throws Exception {
		if (hasNoFilters()){
			listener.tag(name, line);
		}
		else{
			metaBuffer.add(new Sexp(Sexp.Events.TAG, name, line));
		}
		replayBuffersIfAllOk();
	}
	
	@Override
	public void comment(String content, int line) throws Exception {
		if (hasNoFilters()){
			listener.comment(content, line);
		}
		else{
			metaBuffer.add(new Sexp(Sexp.Events.COMMENT, content, line));
		}
		replayBuffersIfAllOk();
	}

	@Override
	public void feature(String keyword, String name, int line) throws Exception {
		System.out.println("Feature Line: " + line);
		if (hasNoFilters()){
			listener.feature(keyword, name, line);
		}
		else{
			Sexp sexp = new Sexp(Sexp.Events.FEATURE, keyword, name, line);
			featureBuffer.clear();
			featureBuffer.addAll(metaBuffer);
			featureBuffer.add(sexp);
			featureTags = extractTags();
			metaBuffer.clear();
			if (filterMatch(sexp)){
				featureOk = true;
			}
		}
		replayBuffersIfAllOk();
	}
	
	@Override
	public void background(String keyword, String name, int line)
			throws Exception {
		if (hasNoFilters()){
			listener.background(keyword, name, line);
		}
		else{
			Sexp sexp = new Sexp(Sexp.Events.BACKGROUND, keyword, name, line);
			featureBuffer.addAll(metaBuffer);
			featureBuffer.add(sexp);
			metaBuffer.clear();
			tableState = TableState.BACKGROUND;
			if (filterMatch(sexp)){
				featureOk = true;
			}
		}
		replayBuffersIfAllOk();
	}

	@Override
	public void scenario(String keyword, String name, int line)
			throws Exception {
		System.out.println("Scenario: " + keyword + ", " + name + "," + line);
		if (hasNoFilters()){
			listener.scenario(keyword, name, line);
		}
		else{
			Sexp sexp = new Sexp(Sexp.Events.SCENARIO, keyword, name, line);
			replayExamplesRowsBuffer();
			scenarioBuffer.clear();
			scenarioBuffer.addAll(metaBuffer);
			scenarioBuffer.add(sexp);
			scenarioTags = extractTags();
			exampleTags.clear();
			metaBuffer.clear();
			if (filterMatch(scenarioBuffer) || tagMatch()){
				scenarioOk = true;
			}
			examplesOk = false;
			tableState = TableState.STEP;
		}
		replayBuffersIfAllOk();
	}

	@Override
	public void scenario_outline(String keyword, String name, int line)
			throws Exception {
		if (hasNoFilters()){
			listener.scenario_outline(keyword, name, line);
		}
		else{
			Sexp sexp = new Sexp(Sexp.Events.SCENARIO_OUTLINE, keyword, name, line);
			replayExamplesRowsBuffer();
			scenarioBuffer.clear();
			scenarioBuffer.addAll(metaBuffer);
			scenarioBuffer.add(sexp);
			scenarioTags = extractTags();
			exampleTags.clear();
			metaBuffer.clear();
			if (filterMatch(scenarioBuffer)){
				scenarioOk = true;
			}
			examplesOk = false;
			tableState = TableState.STEP;
		}
		replayBuffersIfAllOk();
	}

	@Override
	public void examples(String keyword, String name, int line)
			throws Exception {
		if (hasNoFilters()){
			listener.examples(keyword, name, line);
		}
		else{
			Sexp sexp = new Sexp(Sexp.Events.EXAMPLES, keyword, name, line);
			replayExamplesRowsBuffer();
			examplesBuffer.clear();
			examplesBuffer.addAll(metaBuffer);
			examplesBuffer.add(sexp);
			exampleTags = extractTags();
			metaBuffer.clear();
			examplesRowsBuffer.clear();
			if (filterMatch(examplesBuffer) || tagMatch()){
				examplesOk = true;
			}
			tableState = TableState.EXAMPLES;
		}
		replayBuffersIfAllOk();
	}

	@Override
	public void step(String keyword, String name, int line) throws IOException,
			Exception {
		System.out.printf("Step %s, %s, %d\n", keyword, name, line);
		if (hasNoFilters()){
			listener.step(keyword, name, line);
		}
		else{
			Sexp sexp = new Sexp(Sexp.Events.STEP, keyword, name, line);
			if (tableState.equals(TableState.BACKGROUND)){
				featureBuffer.addAll(metaBuffer);
				featureBuffer.add(sexp);
				metaBuffer.clear();
				if (filterMatch(sexp)){
					featureOk = true;
				}
			}
			else{
				scenarioBuffer.add(sexp);
				scenarioOk |= filterMatch(scenarioBuffer);
				tableState = TableState.STEP;
			}
		}
		replayBuffersIfAllOk();
	}

	@Override
	public void row(List<String> row, int line) throws Exception {
		if (hasNoFilters()){
			listener.row(row, line);
		}
		else{
			Sexp sexp = new Sexp(Sexp.Events.ROW, row, line);
			if (tableState.equals(TableState.EXAMPLES)){
				if (!headerRowAllreadyBuffered()){
					examplesBuffer.add(sexp);
					if (filterMatch(examplesBuffer)){
						examplesOk = true;
					}
				}
				else{
					if (scenarioOk || examplesOk || featureOk || filterMatch(sexp)){
						examplesRowsBuffer.add(sexp);
					}
				}
			}
			else if (tableState.equals(TableState.STEP)){
				scenarioBuffer.add(sexp);
				scenarioOk |= filterMatch(scenarioBuffer);
			}
			else if (tableState.equals(TableState.BACKGROUND)){
				featureBuffer.addAll(metaBuffer);
				featureBuffer.add(sexp);
				metaBuffer.clear();
			}
			else{
				throw new RuntimeException("Bad table_state:" + tableState);
			}
		}
		replayBuffersIfAllOk();
	}

	@Override
	public void py_string(String string, int line) throws Exception {
		if (hasNoFilters()){
			listener.py_string(string, line);
		}
		else{
			scenarioBuffer.add(new Sexp(Sexp.Events.PY_STRING, string, line));
			scenarioOk |= filterMatch(scenarioBuffer);
		}
		replayBuffersIfAllOk();
	}

	@Override
	public void eof() throws Exception {
		replayExamplesRowsBuffer();
		listener.eof();
	}

	@Override
	public void syntax_error(String state, String event,
			List<String> legalEvents, int line) throws Exception {
	}

	private void detectFilters(List filters) {
		if (hasNoFilters()){
			return;
		}
		checkIfMoreThanOneFilterType(filters);
		this.filterMethod = new FilterMethodFactory().getFilterMethod(filters.get(0).getClass());
		this.filterMethod.setFilters(filters);
	}

	private void checkIfMoreThanOneFilterType(List filters) {
		Set<Class> filterTypes = new HashSet<Class>();
		for (Object object : filters) {
			filterTypes.add(object.getClass());
		}
		if (filterTypes.size() > 1){
			throw new RuntimeException("Bad Filter: " + filterTypes);
		}
	}

	private boolean hasNoFilters() {
		return filters.size() == 0;
	}

	private boolean headerRowAllreadyBuffered() {
		if (examplesBuffer.isEmpty()){
			return false;
		}
		return examplesBuffer.get(0).getEvent() == Sexp.Events.ROW;
	}

	private boolean filterMatch(Sexp sexp) {
		return filterMatch(Arrays.asList(sexp));
	}

	private boolean filterMatch(List<Sexp> buffer) {
		for (Sexp sexp : buffer) {
			if (filterMethod.filter(sexp)){
				return true;
			}
		}
		return false;
	}

	private boolean tagMatch() {
		if (filterMethod.getClass() != TagFilterMethod.class){
			return false;
		}
		return filterMethod.filterTags(currentTags());
	}

	private void replayBuffers() throws Exception {
		List<Sexp> allItems = new ArrayList<Sexp>();
		allItems.addAll(featureBuffer);
		allItems.addAll(scenarioBuffer);
		for (Sexp item : allItems) {
			System.out.println("Replay: " + item);
			item.replay(listener);
		}
		featureBuffer.clear();
		scenarioBuffer.clear();
	}

	private List<String> extractTags() {
		List<String> retVal = new ArrayList<String>();
		for (Sexp sexp : metaBuffer) {
			if (sexp.getEvent() == Sexp.Events.TAG){;
				retVal.add(sexp.getKeyword());
			}
		}
		return retVal;
	}

	private void replayExamplesRowsBuffer() throws Exception {
		if (!examplesRowsBuffer.isEmpty()){
			replayBuffers();
			List<Sexp> exampleItems = new ArrayList<Sexp>();
			exampleItems.addAll(examplesBuffer);
			exampleItems.addAll(examplesRowsBuffer);
			for (Sexp sexp : exampleItems) {
				System.out.println("replayExamplesRowsBuffer: " + sexp);
				sexp.replay(listener);
			}
			examplesRowsBuffer.clear();
		}
	}

	private void replayBuffersIfAllOk() throws Exception {
		if (scenarioOk || examplesOk || featureOk){
			System.out.println("replayBuffersIfAllOk");
			replayBuffers();
		}
	}

	private List currentTags() {
		List retVal = new ArrayList();
		retVal.addAll(featureTags);
		retVal.addAll(scenarioTags);
		retVal.addAll(exampleTags);
		return retVal;
	}

}
